cmd/modelcmd/controller.go

```diff
@@ -75,28 +88,73 @@ func (c *ControllerCommandBase) SetClientStore(store jujuclient.ClientStore) {
 
 // ClientStore implements the ControllerCommand interface.
 func (c *ControllerCommandBase) ClientStore() jujuclient.ClientStore {
+	c.assertRunStarted()
 	return c.store
 }
 
-// SetControllerName implements ControllerCommand.SetControllerName.
-func (c *ControllerCommandBase) SetControllerName(controllerName string, allowDefault bool) error {
+func (c *ControllerCommandBase) initController() error {
```

Comment:
I agree that it would be good to move towards a more composition-based model. FWIW I started by making it avoid calling Run if the context couldn't be initialised, but then did this lazy implementation when I realised just how many tests rely on being able to call Run without any models or controllers in the store.